R/functions_degs.R

```diff
@@ -106,37 +157,509 @@ DegsWriteToFile = function(degs, annot_ensembl, file) {
+    
+    if (!is.na(contrast[["subset_column"]])) {
+      valid = TRUE
+      
+      # subset_column but not subset_group specified
+      if (is.na(contrast[["subset_column"]])) {
+        error_messages = c(error_messages, paste("The 'subset_column' column must be used together with the 'subset_group' column (in row", i ,"of the deg contrasts table)."))
+        valid = FALSE
+      }
+      
+      # subset_column is not a column of cell metadata
+      if (valid && !contrast[["subset_column"]] %in% colnames(cell_metadata)) {
+        error_messages = c(error_messages, paste0("The 'subset_column' column value '",contrast[["subset_column"]], "' in row ", i, " of the deg contrasts table is not part of the cell metadata."))
+        valid = FALSE
+      }
+      
+      if (valid) {
+        if (nchar(contrast[["subset_group"]])==0) {
+          # subset_group is empty: get all levels or unique values and join them by semicolon 
+          subset_group_values = levels(cell_metadata[, contrast[["subset_column"]], drop=TRUE])
+          if (length(subset_group_values)==0) subset_group_values = unique(sort(cell_metadata[, contrast[["subset_column"]], drop=TRUE])) 
+          contrast[["subset_group"]] = JoinSpecificationList(subset_group_values, first_level_separator=";")
+        } else {
+          # subset_group is not empty: split by semicolon, then by plus; then validate values 
+          subset_group = SplitSpecificationString(contrast[["subset_group"]], first_level_separator=";", second_level_separator="+")
+          if (any(!unlist(subset_group) %in% unique(cell_metadata[, contrast[["subset_column"]], drop=TRUE]))) {
+            error_messages = c(error_messages, paste("At least one value of column 'subset_group' in row",i,"of the deg contrasts table cannot be found in the column '", contrast[["subset_column"]] , "' of the cell metadata."))
+          }
+          contrast[["subset_group"]] = JoinSpecificationList(subset_group, first_level_separator=";", second_level_separator="+")
+        }
+      }
+    }
+    
+    # test
+    if (!"test" %in% names(contrast) || is.na(contrast[["test"]])) contrast[["test"]] = "wilcox"
+    if (!contrast[["test"]] %in% c("wilcox", "bimod", "roc", "t", "negbinom", "poisson", "LR", "MAST", "DESeq2")) error_messages = c(error_messages, paste0("The 'test' column value '",contrast[["test"]], "' in row ", i, " of the deg contrasts table must be one of: 'wilcox', 'bimod', 'roc', 't', 'negbinom', 'poisson', 'LR', 'MAST', 'DESeq2'."))
+    
+    # padj
+    if (!"padj" %in% names(contrast) || is.na(contrast[["padj"]])) contrast[["padj"]] = 0.05
+    
+    # log2FC
+    if (!"log2FC" %in% names(contrast) || is.na(contrast[["log2FC"]])) contrast[["log2FC"]] = 0
+    
+    # min_pct
+    if (!"min_pct" %in% names(contrast) || is.na(contrast[["min_pct"]])) contrast[["min_pct"]] = 0.1
+    
+    # assay
+    if (!"assay" %in% names(contrast) || is.na(contrast[["assay"]])) contrast[["assay"]] = "RNA"
```

Comment:
Depends... . On the one hand, I think most tests want the raw/at least normalised data which for us is `RNA`. But it could also be set to the default of sc.

Up to you.

Reply:
Nah you are right.